README.md
# museduino
an open source electronics environment for exhibits
| <a href="https://museduino.org">museduino.org</a>


## PCB design files and Schematics

All files are under <a href="https://github.com/CTDL/museduino/tree/master/hardware">hardware</a>  folder and subdivided by PCB design software.

ExpressPCB folder contains versions 2.2 and below. Kicad folder is organized by board name contains its own versions (3.0 or greater).

View changelog.md for further information on changes between versions.


## Tutorials and Circuit Diagrams
Content for workshops and museduino guides can be found under 
<a href="https://github.com/CTDL/museduino/tree/master/tutorials">tutorials</a> folder.


## Solder Instructions
Component guides and step-by-step instructions coming soon.
<a href="https://github.com/CTDL/museduino/tree/master/solder-instructions">solder-instructions</a> folder.


hardware/CHANGELOG.md
# v4.0.3 (Dec 2019)

Fixed main shield labels on Analog select. 

# v4.0.2 (July 2018)

Kicad 5 update. Adjusted main shield labels.

Additional Satellite board named Port Extender is designed for projects that need to control more input and output devices.

#### Port Extender v1.0 
The Port Extender board was designed for use with the dedicated i2c port, Port D on the Main Shield and intended for projects that need to control more input and output devices. The Port Extender is equipped with two MCP23017 i2c 16-bit i/o digital port expanders, a CD74HC4051 8-channel multiplexer, and a CD74HC4067 16-channel multiplexer.  The Port Extender uses i2c pins (A4/A5 on Arduino Uno), analog pin (A3), and a PWM pin (D5~) in order to increase the number of usable digital, analog, and PWM pins.

# v4.0 (May 2018)

Adjusted main shield pin configuration. 

##### Analog Select on Main Shield

The Museduino main shield now allows distributed Analog pins to switch between ports. For example, pin D11~ on port B can switch places with pin A0 on port A. Pin A3 on Port C can also switch with Pin D2 on port D. This feature is useful when you need an extra Analog pin to be accessed from a different port. Intended for use with optional Port Extender.


##### New Distrubuted I/O

The Museduino shield distributes 20 Arduino pins between 4 CAT5/RJ-45 “ports”. A Satellite board connected on either port allows access to 5 specific Arduino pins to be used for sensors/actuators. The shield and satellite boards have a labeled solder mask to assist with configuration. Ports are labeled with a circled A, B, C, or D on the shield. The 5 Satellite I/O pins are labeled with a circled number 1 through 5 on the Satellite boards.

|  | A | B | C | D |	
| ------------- | ------------- | ------------- | ------------- | ------------- |
| 1 | A0  | D13  | * A3 | A4 |
| 2 | A1  | D12  | A2 | A5 |
| 3 | * D8  | D11~  | ** D5~ | * D2 |
| 4  | D7  | D10~  | D4 | D1 |
| 5  | D6~  | * D9~  | D3~ | ** D0 |

~ arduino pwm pin
* pwm switch

** analog switch


# v3 (Jan 2018)

All boards redesigned with Kicad 4. No functionaly changed, but replaced DPDT switch on main shield due to cost increase.  

Printed small run for testing purposes. 

# v2.2 (June 2015) 

Designed with Express PCB

No changes to schematic. Improved parts layout on pcb. 


# v2.0 (June 2015)

Designed with Express PCB. 

All i/o on satellite boards now use 4 pin screw terminals instead of header pins.

Improvements to overall pcb label design. Ports and satellite i/o labels use circled letter or number to match charts in documentation. 


##### New Distrubuted I/O labels

The Museduino distributes 5 different Arduino pins on ports A, B, C, or D labeled on the shield. A satellite connected on either port allows access to pins labeled 1, 2, 3, 4, or 5.

| ports | A | B | C | D |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| 1 | A0  | D13  | A2 | A4 |
| 2 | A1  | D12  | A3 | A5 |
| 3 | * D8  | D11~  | D5~ | * D2 |
| 4 | D7  | D10~  | D4 | D1 |
| 5 | D6~  | * D9~  | * D3~ | D0 |

~ arduino pwm pin
* pwm switch


# v1.7 (May 2015)
Designed with Express PCB

Simplified labelling for Distributed I/O on all boards and chart using letters and numbers.

##### New Distrubuted I/O chart

The Museduino distributes 5 different Arduino pins on ports A, B, C, or D labeled on the shield. A satellite (S) connected on either port allows access to arduino pins labeled 1, 2, 3, 4, or 5.

| s | A | B | C | D |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| 1 | A0  | D13  | A2 | A4 |
| 2 | A1  | D12  | A3 | A5 |
| 3 | * D8  | D11~  | D5~ | * D2 |
| 4 | D7  | D10~  | D4 | D1 |
| 5 | D6~  | * D9~  | * D3~ | D0 |

~ arduino pwm pin
* pwm switch


# v1.5 (March 2015)

Designed with Express PCB

New labels on PCB and to match Distributed I/O chart.

Main shield allows distributed PWM pins to switch between ports (cat5 pinout). For example, pin D9~ on port B can switch places with pin D8 on port A. Pin D3~ on Port D can also switch with Pin D2 on port C. This feature is useful when you need an extra PWM pin to be accessed from a Satellite board.

#### Smorgasboard 
Smorgasboard satellite board has 5 sets of headers for connecting to sensors/actuators labeled A through E in the schematic. Each i/o has its own pin configuration. 

A, B - power, signal pin, signal pin, ground pin
C - power, signal with open drain to ground (mosfet), direct signal, ground pin
D, E - power, signal pin with resistor n series, signal pin, ground pin

##### External Power board 
Additional Satellite board named External Power board is designed for sensor and actuators that require their own external power supply (NeoPixels, DC Motors, etc). Since satellites connected via CAT5 cables can be as long or short as needed, this board is particularly useful for large-scale installations.

External Power satellite board has 5 sets of headers for connecting to sensors/actuators labeled A through E in the schematic. Each i/o has its own pin configuration.  

A, B - power, signal pin, signal pin, ground pin
C, E - power, signal with open drain to ground (mosfet), direct signal, ground pin
D - power, signal pin, signal pin with resistor n series, ground pin

C, D and E contains a jumper that allows you to switch voltage between 5v from cat5, or voltage from the external power source. 

##### New Distrubuted I/O chart

The Museduino distributes 5 different Arduino pins on ports A, B, C, or D labeled on the shield. A satellite (S) connected on either port allows access to the specified Arduino pins used for sensors/actuators. 

| s  | s | A | B | C | D |
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| smorgasboard pins  |  ext power pins | j5 | j6 | j8 | j7 |	
| j14 | j19 | A0  | D13  | A2 | A4 |
| j16  | j20 | A1  | D12  | A3 | A5 |
| j12  | j21 | * D8  | D11~  | D5~ | * D2 |

tutorials/PortExtender-circuitpython/PortExtenderTutorial.py
# Museduino Port Extender Tutorial

'''
Arduino Uno pins A4 and A5 are technically tied to SCL and SDA.
Circuit python boards with the Arduino footprint do not always use the same i2c pins.
You may be able to substitute the pins via code like below:
'''
#import busio
#i2c = busio.I2C(board.A4, board.A5)

import Museduino_PortExtender
portExtender = Museduino_PortExtender()

# access mcp0 or mcp1 chip separately
mcp0 = portExtender.mcp0
digitalio = portExtender.digitalio

# mcp digital output
led = mcp0.get_pin(5)
led.direction = digitalio.Direction.OUTPUT

# mcp digital input with a pull-up resistor enabled
btn = mcp0.get_pin(8)
btn.direction = digitalio.Direction.INPUT
btn.pull = digitalio.Pull.UP

# analog channel
potpin = 0

# pwm channel
pwm_led = 6

while True:

    if btn.value:
        led.value = True
    else:
        led.value = False

    potval = portExtender.readAnalog(potpin)
    portExtender.writePWM(potval)

tutorials/PortExtender-circuitpython/Museduino_PortExtender.py
# Museduino Port Extender class (place in libs folder)

import board
import busio
import digitalio
import analogio
import pulseio
import adafruit_mcp230xx

class Museduino_PortExtender:

    # Initialize the I2C bus:
    i2c = busio.I2C(board.SCL, board.SDA)

    # Initialize the MCP23017 chips:
    # U1 address 000
    mcp0 = adafruit_mcp230xx.MCP23017(i2c, address=0x20)
    # U2 address 001
    mcp1 = adafruit_mcp230xx.MCP23017(i2c, 0x21)

    # Analog channels 0 - 15
    analog_common_input = analogio.AnalogIn(board.A3)  # A3
    analog_controlPins = [9, 10, 11, 12]  # s0, s1, s2, s3

    # PWM channels 0 - 8
    pwm_common_output = pulseio.PWMOut(board.D5)  # D5
    pwm_controlPins = [13, 14, 15]  # s0, s1, s2

    def __init__(self):

        # Initialize control pins for multiplexers

        for i in range(4):
            pin = self.mcp1.get_pin(self.analog_controlPins[i])
            pin.direction = digitalio.Direction.OUTPUT
            pin.pull = digitalio.Pull.UP
            pin.value = False

        for i in range(3):
            pin = self.mcp1.get_pin(self.pwm_controlPins[i])
            pin.direction = digitalio.Direction.OUTPUT
            pin.pull = digitalio.Pull.UP
            pin.value = False

    # Analog Input
    def readAnalog(self, chan):
        for i in range(4):
            pin = self.mcp1.get_pin(self.analog_controlPins[i])
            pin.pull = digitalio.Pull.UP
            pin.direction = digitalio.Direction.OUTPUT
            pin.value = bool(chan & 1)

            # shift bit
            chan = chan >> 1

        val = self.get_voltage(self.analog_common_input)
        return val

    # helper function
    def __get_voltage(pin):
        return (pin.value * 3.3) / 65536

    # PWM output
    def writePWM(self, chan, val):
        for i in range(3):
            pin = self.mcp1.get_pin(self.pwm_controlPins[i])
            pin.direction = digitalio.Direction.OUTPUT
            pin.pull = digitalio.Pull.UP
            pin.value = bool(chan & 1)

            # shift bit
            chan = chan >> 1

        self.pwm_common_output.duty_cycle = val

tutorials/Museduino_PortExtender_Tutorial/Museduino_PortExtender.h
#include "Wire.h"
#include "Adafruit_MCP23017.h" // MCP23017 I/O expander Library

Adafruit_MCP23017 mcp0; //U1 on Port Extender board
Adafruit_MCP23017 mcp1; //U2 on Port Extender board

///////////////////////////////////////
// ANALOG MUX - CD74HC4067
// U3 on Port Extender board
// 16 analog input pins - see analog_mux_channel below and chart on PCB
// 

//Requires 4 mcp1 library pins and common input from Arduino - see Analog jumpers on Museduino main shield
int analog_s0 = 9;
int analog_s1 = 10;
int analog_s2 = 11;
int analog_s3 = 12;
int analog_common_input = A3; // analog mux common input from arduino
int analog_controlPin[] = {analog_s0, analog_s1, analog_s2, analog_s3}; 

//analog mux channels for additional analog pins 0-15 
int analog_mux_channel[16][4]={ 
// s0,s1,s2,s3  channel
  {0,0,0,0}, // 0 
  {1,0,0,0}, // 1 
  {0,1,0,0}, // 2 
  {1,1,0,0}, // 3 
  {0,0,1,0}, // 4 
  {1,0,1,0}, // 5 
  {0,1,1,0}, // 6 
  {1,1,1,0}, // 7 
  {0,0,0,1}, // 8 
  {1,0,0,1}, // 9 
  {0,1,0,1}, // 10 
  {1,1,0,1}, // 11 
  {0,0,1,1}, // 12 
  {1,0,1,1}, // 13 
  {0,1,1,1}, // 14 
  {1,1,1,1}  // 15 
};

///////////////////////////////////////
// PWM MUX - CD74HC4051E
// U4 on Port Extender board
// 8 pwm output pins - see pwm_mux_channel below and chart on PCB

//Requires 3 mcp1 library pins for and common output from Arduino - see PWM jumpers on Museduino main shield
int pwm_s0 = 13;
int pwm_s1 = 14;
int pwm_s2 = 15;
int pwm_common_output = 5; // mux common output from Arduino 
int pwm_controlPin[] = {pwm_s0, pwm_s1, pwm_s2}; 

//pwm mux channels for additional pwm pins 0-7
int pwm_mux_channel[8][3]={ 
// s0,s1,s2,s3  channel
  {0,0,0}, // 0 
  {1,0,0}, // 1 
  {0,1,0}, // 2 
  {1,1,0}, // 3 
  {0,0,1}, // 4 
  {1,0,1}, // 5 
  {0,1,1}, // 6   
  {1,1,1}, // 7 
};

void portextender_setup() {

  //MCP0 - see MCP23017 datasheet and library for address selection
  //default is 0
  mcp0.begin(0); // Address 000 (jmp1 - GND, jmp2 - GND, jmp3 - GND)

  //MCP1
  mcp1.begin(1); //Address 001 (jmp4 - POWER, jmp5 - GND, jmp6 - GND)

  //Analog Mux INPUT control pins from MCP1
  for(int i = 0; i < 4; i ++){
    mcp1.pinMode(analog_controlPin[i], OUTPUT); 
    mcp1.pullUp(analog_controlPin[i], HIGH);
    mcp1.digitalWrite(analog_controlPin[i], LOW); 
  } 
  pinMode(analog_common_input, INPUT);

  //PWM Mux OUTPUT control pins from MCP1
  for(int i = 0; i < 3; i ++){
    mcp1.pinMode(pwm_controlPin[i], OUTPUT); 
    mcp1.pullUp(pwm_controlPin[i], HIGH);
    mcp1.digitalWrite(pwm_controlPin[i], LOW); 
  } 
  pinMode(pwm_common_output, OUTPUT);
  
}

//ANALOG MUX - read function for Analog Input
int readAnalogMux(int chan){  

  Serial.print("analog chan: ");
  
 //loop through the 4 sig pins control pins
  for(int i = 0; i < 4; i ++){
    Serial.print(analog_mux_channel[chan][i]);
    mcp1.digitalWrite(analog_controlPin[i], analog_mux_channel[chan][i]); 
  } 
  Serial.println();

  //read the value
  int val = analogRead(analog_common_input); //return the value 
  return val; 
  
 }

 //PWM MUX - write function for PWM Output
int writeAnalogPwmMux(int chan, int value){  

    Serial.print("pwm chan: ");
    
    //loop through the 3 sig control pins
    for(int i = 0; i < 3; i ++){
      Serial.print(pwm_mux_channel[chan][i]);
      mcp1.digitalWrite(pwm_controlPin[i], pwm_mux_channel[chan][i]); 
    } 
    Serial.println();

    Serial.print("pwm mux value: ");
    Serial.println(value);

    //write the value
    analogWrite(pwm_common_output,value); //return the value 
    
 }

//PWM MUX - write function for Digital Output
int writeDigitalPwmMux(int chan, int value){  

    Serial.print("pwm chan: ");
    
    //loop through the 3 sig control pins
    for(int i = 0; i < 3; i ++){
      Serial.print(pwm_mux_channel[chan][i]);
      mcp1.digitalWrite(pwm_controlPin[i], pwm_mux_channel[chan][i]); 
    } 
    Serial.println();

    Serial.print("pwm mux value: ");
    Serial.println(value);

    //write the value
    digitalWrite(pwm_common_output,value); //return the value 
    
 }
﻿
